lib/ProductOpener/Data.pm

```diff
@@ -60,45 +60,58 @@ sub execute_query {
 	my ($sub) = @_;
 
 	return Action::Retry->new(
-          attempt_code => sub { $action->run($sub) },
-          on_failure_code => sub { my ($error, $h) = @_; die $error; }, # by default Action::Retry would return undef
-		  # If we didn't get results from MongoDB, the server is probably overloaded
-		  # Do not retry the query, as it will make things worse
-		  strategy => { Fibonacci => { max_retries_number => 0, } },
-      )->run();
+		attempt_code => sub { $action->run($sub) },
+		on_failure_code => sub { my ($error, $h) = @_; die $error; }, # by default Action::Retry would return undef
+		# If we didn't get results from MongoDB, the server is probably overloaded
+		# Do not retry the query, as it will make things worse
+		strategy => { Fibonacci => { max_retries_number => 0, } },
+	)->run();
 }
 
 sub get_products_collection {
-	return get_collection($mongodb, 'products');
+	my ($timeout) = @_;
+	return get_collection($mongodb, 'products', $timeout);
 }
 
 sub get_products_tags_collection {
-	return get_collection($mongodb, 'products_tags');
+	my ($timeout) = @_;
+	return get_collection($mongodb, 'products_tags', $timeout);
 }
 
 sub get_emb_codes_collection {
-	return get_collection($mongodb, 'emb_codes');
+	my ($timeout) = @_;
+	return get_collection($mongodb, 'emb_codes', $timeout);
 }
 
 sub get_recent_changes_collection {
-	return get_collection($mongodb, 'recent_changes');
+	my ($timeout) = @_;
+	return get_collection($mongodb, 'recent_changes', $timeout);
 }
 
 sub get_collection {
-	my ($database, $collection) = @_;
-	return get_mongodb_client()->get_database($database)->get_collection($collection);
+	my ($database, $collection, $timeout) = @_;
+	return get_mongodb_client($timeout)->get_database($database)->get_collection($collection);
 }
 
 sub get_database {
-	return get_mongodb_client()->get_database($mongodb);
+	my ($database) = @_ // $mongodb;
+	return get_mongodb_client()->get_database($database);
 }
 
 sub get_mongodb_client() {
+	my ($timeout) = @_;
+
+	my %client_options = (
+		host => $mongodb_host,
+
+		# https://metacpan.org/pod/MongoDB::MongoClient#socket_timeout_ms
+		# default is 30000 ms
+		socket_timeout_ms => $timeout // 30000,
+	);
+
 	if (!defined($client)) {
```

Comment:
Probably not an issue with the `.pl` script, but since the client will be cached in https://github.com/openfoodfacts/openfoodfacts-server/blob/a7f89889597d6cec7f9b088b362b647b95f7d24f/lib/ProductOpener/Data.pm#L56, calling `get_mongodb_client` with a different timeout than in the first call would return the original `$client` with the original timeout. Maybe we'd need to keep a hash of clients with the timeout as they key? I'm not sure. :slightly_smiling_face:

Reply:
i think having a timeout for the web site is probably fine. 
although i was reading something that suggested having multiple connections, so the site isn't queueing all queries into one connection, which might be part of why mongodb stuff is slow, i.e. if other requests are waiting behind a slow query.

Reply:
of course, mongodb is also just slow in general.

Reply:
There is no queueing of multiple MongoDB queries into one connection today. Each of the mod_perl process has one MongoDB client and connection, each serves only one web query at a time, and each of those web queries makes mongodb queries in sequence.

Reply:
The potentital issue isn't about queueing (which might be useful, but as the client isn't used in parallel (on the web), I'm not sure if it helps). It's about
```Perl
my $client500 = get_mongodb_client(500);
my $client5000 = get_mongodb_client(5000);
print "yes" if $client500 == $client5000; // Probably prints yes
```
Because only one `$client` is created.

Reply:
well, re stephane's comment below, the use case I was trying to solve is 1, the update_all_products/export_database type stuff, which regularly dies at the slightest mongodb wobble. not mod_perl web pages specifying their own timeouts, so that gotcha could just be explained in a comment in the function. but i like the idea of use case 2 as well.